### PR TITLE
Implement ValidatorRule and refactor Validator

### PR DESCRIPTION
**Pull request by mkrecek234**

### CODE REVIEW

**Comment by mvorisek** [comment]
CI enabled, repush. Then much more work is before you as outlined in the chat :)

**Comment by mvorisek** [comment]
@abbadon1334 I analysed more places to fix a few days ago with @mkrecek234. Did it get to you:
> some repo files seem outdated, do a diff against atk4/ui and atk4/data repos and unify the CI structure as much as possible, for ex. the phpstan should not be configured with treatPhpDocTypesAsCertain: false, there should be no tools/CoverageUtil.php (use https://github.com/atk4/ui/blob/develop/src/Behat/CoverageUtil.php) etc.

for ex. the tests should use `self::assert*`, I wonder if the CI is fully working (CS Fixer itself is probably too short as there is intermediate class)

File: .github/workflows/test-unit.yml [hunk 1]
```diff
@@ -24,20 +24,20 @@ jobs:
...
       - name: Checkout
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
 
       - name: Configure PHP
         run: |
           rm /usr/local/etc/php/conf.d/docker-php-ext-xdebug.ini
           php --version
-
+
```
Comment (mvorisek) [inline .github/workflows/test-unit.yml]: fix trailing ws

File: composer.json [hunk 2]
```diff
@@ -30,13 +30,13 @@
...
-    "php": ">=7.4 <8.2",
+    "php": ">=7.4 <8.3",
     "ext-pdo": "*",
     "atk4/data": "dev-develop",
     "vlucas/valitron": "^1.4"
   },
   "require-release": {
-    "php": ">=7.4 <8.2",
+    "php": ">=7.4 <8.3",
     "atk4/data": "~3.1.0",
```
Comment (mvorisek) [inline composer.json]: this is definitely not up-to-date too

**Comment by abbadon1334** [comment]
@mvorisek @mkrecek234 i have done some changes, can you check it again?

File: src/ValidatorRule.php [hunk 3]
```diff
@@ -0,0 +1,131 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Atk4\Validate;
+
+use Atk4\Data\Exception;
+use Atk4\Data\Model;
+
+class ValidatorRule
```
Comment (mvorisek) [inline src/ValidatorRule.php:10]: This is probably not needed to align /w atk4, thus it should be not part of this PR.

the correct way would be:
- one PR to fix CS/stan (for atk4/ui 4.0)
- do 4.0 release
- then PR to align /w atk4/ui 5.0
- if refactoring is wanted, one or more isolated PRs
- test manually extensively as I doubt everything is tested using CI in this repo
- do 5.0 release

Comment (abbadon1334) [reply src/ValidatorRule.php:10]: This lib allow bridging an easy with valitron.
we need Atk4/Data in __construct to link Validate hook and in validate method we have Atk4\Data\Model as first argument.
Actually use very simple methods:
`$model->onHook`
`$model->get()`
`$model->get($fieldName)`

i create a PR that fix CS but still have a problem with stan : #13 

But probably due to simple use of Model we can allow dependency of Atk4/Data > 3.1.0 but in this way we need to test against 3.1 4.0 5.0 or thinking this is a marginal repo, do only validation and use really solid and low level methods of Model, we can probably leave as it is WDYT?

Comment (mvorisek) [reply src/ValidatorRule.php:10]: if (upcoming) rel 4.0 will pass all tests /w atk4/data 3.1, we can allow both (3.1 + 4.0)

File: demos/_demo-data/create-db.php [hunk 4]
```diff
@@ -19,12 +19,13 @@
...
-if (getenv('GITHUB_JOB') === 'unit-test') {
+if (getenv('GITHUB_JOB') !== false && strpos(getenv('GITHUB_JOB'), 'unit-') === 0) {
     echo "skip db creation in create-db\n\n";
 
     return;
 }
 
-(new Migrator(new Dummy($db)))->create();
+$migr = new Migrator(new Dummy($db));
+$migr->dropIfExists()->create();
```
Comment (mvorisek) [inline demos/_demo-data/create-db.php]: this is exactly what we do not want, we want to be sure this skip is called only once on a clean testing DB

**Comment by DarkSide666** [comment]
closed in favor of https://github.com/atk4/validate/pull/20

**Pull request closed without merging**